fix(workflowengine): require a web component as operation plugin

blizzz opened the pull request:
* Contributes to fixing https://github.com/nextcloud/server/issues/49909
* Background: https://github.com/nextcloud-libraries/nextcloud-vue/issues/6360
* Talk counterpart in https://github.com/nextcloud/spreed/pull/14367

## Summary

solves an incompatibility issue when the providing app registers their code from an incompatible nextcloud-vue version.

Also changes and clarifies WorkflowEngine API. This is necessary to stay compatible with the original way, but also promotes usage of the originally declared but never used "component" attribute on registration.

## TODO

- [ ] ...

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [ ] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [ ] Screenshots before/after for front-end changes
- [ ] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [ ] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

blizzz committed:
fix(workflowengine): require a web component as operation plugin

solves an incompatibility issue when the providing app registers their code
from an incompatible nextcloud-vue version.

Also changes and clarifies WorkflowEngine API. This is necessary to stay
compatible with the original way, but also promotes usage of the originally
declared but never used "component" attribute on registration.

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz commented:
/backport to stable30

blizzz commented:
Might need to do the same stuff for registering `Checks`, need to verify.

susnux commented on apps/workflowengine/src/components/Rule.vue:
```diff
@@ -126,10 +132,23 @@ export default {
 		this.originalRule = JSON.parse(JSON.stringify(this.rule))
+
+		if (this.operation?.component) {
+			const View = this.operation.component
+			this.component = new View()
+			this.component.$mount(this.$refs.operationComponent)
+			this.component.$on('input', this.updateOperation)
+			this.component.$props.value = this.rule.operation
```
I prefer the webcomponent approach, but we should be  consistent and either also do so on viewer or use option 2

juliusknorr replied:
I think in viewer we still also use the problematic approach, at least I'm not aware of a register method there:

https://github.com/nextcloud/viewer/blob/42208133a0dab8107c3c1edff73746fc19b2c775/src/models/file.js#L15C1-L15C20
https://github.com/nextcloud/viewer/blob/42208133a0dab8107c3c1edff73746fc19b2c775/src/views/Viewer.vue#L138
https://github.com/nextcloud/viewer/blob/42208133a0dab8107c3c1edff73746fc19b2c775/README.md?plain=1#L148-L177

susnux replied:
@juliusknorr I mean for the next version https://github.com/nextcloud/viewer/issues/2395

max-nextcloud replied:
I'm fine with both - but I agree the Webcomponent looks more elegant.
What would the event listener in the workflow engine look like? Could we do something like:
```
<component :is="'custom-operation'"  @input="handleInput" />
```
or even use `v-model` as we currently do?

max-nextcloud replied:
`@vue/web-component-wrapper` docs say:

> Custom events emitted on the inner Vue component are dispatched on the custom element as a CustomEvent. Additional arguments passed to $emit will be exposed as an Array as event.detail.

Sounds like event handing should work.

blizzz commented:
Thank you all for the input here!

I pushed the suggested changes hereto as well as to Talk's flow. 

I could not get the `:model-value` thing to work, I always ended up with `Property or method "inputValue" is not defined on the instance but referenced during render. `.

Also, the `@update` event did not trigger, got it working with `@input` only – which is also emitted by Talk`s flow. Less changes to existing code the better, it's bad enough that stuff is breaking and maintaining backwards compatibility is a goal as well.  How much of a problem is that?

But I might have been holding it wrong, this is far from my area of confidence :sweat_smile: 

Didn't do yet:
- [ ] satisfying linter
- [ ] squash and rephrase commit
- [ ] update documentation once more

blizzz commented:
> component (componentVue2) - for easy migration for existing apps with Vue.extend

if the app has to be touched anyway – it has – there is no reason to have a bridge solution there (provided WebComponents works). less is more.

max-nextcloud commented on apps/workflowengine/src/components/Rule.vue:
```diff
@@ -29,8 +29,11 @@
-					ref="operationComponent" />
+				<component v-if="operation.component"
+						   :is="operation.component"
+						   @input="updateOperationByEvent"
```
You will need the code from @ShGKme here - the event from the component will be called `update:model-value`, the prop is `model-value`:
```
  :model-value="inputValue"
  @update:model-value="updateOperationByEvent"
```

max-nextcloud replied:
Yes... you need to add it to the data section with a default. Insert line 107, somewhat like this:
```
inputValue: '',
```

blizzz replied:
this fixes the error message at least (i knew it was pebcak)

blizzz replied:
The @update event however does not fire

blizzz replied:
I take your offer for a call :sunglasses:

max-nextcloud commented on apps/workflowengine/src/components/Rule.vue:
```diff
@@ -148,8 +148,12 @@ export default {
+		},
+		async updateOperationByEvent(event) {
+			this.$set(this.rule, 'operation', event.detail[0])
+			this.$refs.operationComponent.value = this.rule.operation
```
I don't think changing the value directly like this is a good idea - not sure if it would work.

Instead I see two options:
1. Use a `data` value that sets the prop:
    ```js
    this.inputValue = $event.detail[0]
    ```
    This matches the code I proposed above.
  
2. use `rule.operation` to set the prop in the template above (instead of `inputValue`). The line above already sets `rule.operation` in a reactive way. So if you use that in the prop you should be done.

blizzz replied:
> I don't think changing the value directly like this is a good idea - not sure if it would work.

It's similar to what we did in the first commit. Only, there we interacted with a native Vue component, not a WebComponent. If I recall correctly, `$props` is not available here. But i might have holding it wrong, happy to try out again. But the issue above is the first blocker.

max-nextcloud replied:
Just saw that `this.rule` is a prop in this component. We really should not alter it but instead emit an event if we want the parent component to change it's value.

blizzz replied:
As discussed with Max back then, we keep the existing tampering with `this.rule` as is for a minimal fix. Bringing to a proper state of the art would then be done later on, an does not require changes on the apps anyway.

max-nextcloud commented on apps/workflowengine/src/components/Rule.vue:
```diff
@@ -186,6 +190,7 @@ export default {
 			if (this.rule.id < 0) {
 				this.$store.dispatch('removeRule', this.rule)
 			} else {
+				this.$refs.operationComponent.value = this.originalRule.operation
```
The same as in `updateOperationByEvent` applies here. So:
1) `this.inputValue = this.originalRule.operation`
2) `this.$set(this.rule, 'operation', this.originalRule.operation)`

max-nextcloud reviewed (commented):
I did not take a look at the entire file - just some had a brief glance. Hope the pointers are useful. Happy to have a call to follow up.

blizzz commented:
Updated, if I recall correctly @max-nextcloud, for a minimal fix we want to stay like this.

I still need to 
* [ ] push the changes to the spreed PR
* [ ] check the Check Plugin, i expect it needs similar changes

blizzz commented:
>  still adjust the Check (maybe an opportunity to implement some random but useful check somewhere 🤔 )

and again i am running against my borders of understanding vue :imp:

max-nextcloud commented on apps/workflowengine/src/components/Rule.vue:
```diff
@@ -123,11 +133,24 @@ export default {
 		this.originalRule = JSON.parse(JSON.stringify(this.rule))
+
+		if (this.operation?.webComponent) {
+			this.$refs.operationComponent.value = this.rule.operation
```
I'm not sure what this does. I would expect `inputValue` to be set instead in order to set the value of the operationComponent via the `model-value` prop.

blizzz replied:
I think a left-over or oversight.

blizzz committed:
fix(workflowengine): require a web component as check plugin

Similar case as with operator plugins (check previous commit). Although we
are not aware of an existign problem, it is there in principle, and
asjusting the API we stay consistent with that one from the operations.

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz committed:
fix(workflowengine): adapt check operator FileSystemTag to use web component

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz committed:
fix(workflowengine): minimally adapt check operator FileMimeType to use web component

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz committed:
fix(workflowengine): use adapted mixin again in FileMimeType

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz committed:
fix(workflowengine): adapt check operator RequestURL to use web component

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz committed:
fix(workflowengine): adapt check operator RequestTime to use web component

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

ShGKme commented on apps/workflowengine/src/components/Checks/file.js:
```diff
@@ -34,7 +37,7 @@ const FileChecks = [
 		name: t('workflowengine', 'File MIME type'),
 		operators: stringOrRegexOperators,
-		component: FileMimeType,
+		webComponent: registerWebComponent(FileMimeType, 'oca-workflowengine-file_mime_type'),
```
Although the approach is called [Web Component](https://developer.mozilla.org/en-US/docs/Web/API/Web_components), in the Web API it is [`custom element`](https://developer.mozilla.org/en-US/docs/Web/API/Web_components/Using_custom_elements).

I'd call it `customElement` or just `element` on the API.

For example, in Vue 3 apps they are defined via `defineCustomElement()` and in native HTML it is `customElements.define()`

ShGKme replied:
Also, we should take into account that the name must be unique

ShGKme commented on apps/workflowengine/src/workflowengine.js:
```diff
@@ -15,12 +15,20 @@ import ShippedChecks from './components/Checks/index.js'
+ *  Will be removed in 03/2028.
  * @property {Function} placeholder - Return a placeholder of no custom component is used
  * @property {Function} validate - validate a check if no custom component is used
+ * @property {string} [webComponent] - A web component id as used in window.customElements.define()`.
+ *  It is expected that the ID is prefixed with the app namespace, e.g. oca-myapp-flow_do_this_operation
+ *  It has to emit the `$update:model-value` event when a value was changed.
+ *  The `model-value` property will be set initially with the rule operation value.
```
> It has to emit the `$update:model-value` event when a value was changed.

Just `update:model-value`, without `$`, isn't it?

blizzz committed:
fix(workflowengine): adapt check operator RequestUserAgent to use web component

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz committed:
fix(workflowengine): adapt check operator RequestUserGroup to use web component

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz committed:
refactor: remove mixing from FileMimeType as it is bad practice

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz committed:
fix: properly set value in FileMimeType

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz commented:
Another by-catch (i.e. broken without my changes): https://github.com/nextcloud/server/issues/51832

blizzz committed:
style(vue): satisfy linter

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

max-nextcloud commented on apps/workflowengine/src/mixins/valueMixin.js:
```diff
@@ -4,32 +4,19 @@
 	data() {
 		return {
-			newValue: '',
+			newValue: [],
```
As a cleanup / follow up:

As far as i can tell this mixin is only used in two components:
Checks/RequestURL.vue 
and
Checks/RequestUserAgent.vue 

They both define the newValue in data as a string. So I think it would make sense to remove this definition here or revert it to string and remove it in the components making use of the mixin (or remove the mixin alltogether and add the watch and methods to the two remaining components.

blizzz replied:
I was told mixins turned to be out bad and should not be used anymore. I had a case where they did not work as they were announced. So, follow up should simply eliminate it and move what is needed to those Checks.

blizzz commented:
/compile amend

blizzz commented:
/backport to stable31

blizzz commented:
/backport to stable30

blizzz committed:
build(workflowengine): compile assets

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz commented:
Failing test not related.
